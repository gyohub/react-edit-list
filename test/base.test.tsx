import React from 'react';
import {cleanup, fireEvent, render, waitFor} from '@testing-library/react';

import ReactEditList, * as REL from 'react-edit-list';

import {schema, data, onLoad} from './data';

describe('base', () => {
    it('basic use', async () => {
        const onLoadFn = jest.fn(onLoad);
        const r = render(<ReactEditList schema={schema} onLoad={onLoadFn} />);
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(1);
        expect(r.container.innerHTML).toMatchSnapshot();

        r.rerender(
            <ReactEditList
                schema={schema}
                onLoad={onLoadFn}
                className='table table-light table-fixed align-middle'
                headClassName='table-dark'
                inputClassName='w-100'
                headers={{product: <p>product</p>}}
                thClassName={{
                    product: 'col-4',
                    type: 'col-3',
                    price: 'col-2',
                    stock: 'col-2',
                    buttons: 'col-1'
                }}
                tdClassName={{
                    product: 'product',
                    type: 'type',
                    price: 'price',
                    stock: 'stock',
                    buttons: 'buttons'
                }}
                btnValidateClassName='btn btn-success p-0 m-0'
                btnDeleteClassName='btn btn-danger py-0 px-1 m-0 mx-1'
                btnCancelClassName='btn btn-secondary py-0 px-1 m-0 mx-1'
            />
        );
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(2);
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });

    it('single custom classes', async () => {
        const onLoadFn = jest.fn(onLoad);
        const r = render(<ReactEditList schema={schema} onLoad={onLoadFn} />);
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(1);
        expect(r.container.innerHTML).toMatchSnapshot();

        r.rerender(
            <ReactEditList
                schema={schema}
                onLoad={onLoadFn}
                thClassName={'thClass'}
                tdClassName={'tdClass'}
                trClassName={'trClass'}
            />
        );
        await waitFor(() => expect(r.getByText(/Desk/)));
        expect(onLoadFn).toBeCalledTimes(2);
        expect(r.container.innerHTML).toMatchSnapshot();
        r.unmount();
    });
});
