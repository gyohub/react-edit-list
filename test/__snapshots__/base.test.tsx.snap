// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base basic use 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base basic use 2`] = `"<table class=\\"table table-light table-fixed align-middle\\"><thead class=\\"table-dark\\"><tr><th class=\\"col-4\\"><p>product</p></th><th class=\\"col-3\\">type</th><th class=\\"col-2\\">price</th><th class=\\"col-2\\">stock</th><th class=\\"col-1\\"></th></tr></thead><tbody><tr><td class=\\"product\\">Desk</td><td class=\\"type\\">capex</td><td class=\\"price\\">100</td><td class=\\"stock\\">20</td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></div></td></tr><tr><td class=\\"product\\">Printer</td><td class=\\"type\\">capex</td><td class=\\"price\\">500</td><td class=\\"stock\\">10</td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></div></td></tr><tr><td class=\\"product\\">Paper</td><td class=\\"type\\">consumable</td><td class=\\"price\\">5</td><td class=\\"stock\\">2000</td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></div></td></tr><tr><td class=\\"product\\">Chair</td><td class=\\"type\\">capex</td><td class=\\"price\\">50</td><td class=\\"stock\\">50</td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></div></td></tr><tr><td class=\\"product\\">Computer</td><td class=\\"type\\">capex</td><td class=\\"price\\">1000</td><td class=\\"stock\\">20</td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></div></td></tr><tr><td class=\\"product\\">Rent</td><td class=\\"type\\"></td><td class=\\"price\\">2000</td><td class=\\"stock\\"></td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn btn-danger py-0 px-1 m-0 mx-1\\">x</button></div></td></tr><tr><td class=\\"product\\">&nbsp;</td><td class=\\"type\\">&nbsp;</td><td class=\\"price\\">&nbsp;</td><td class=\\"stock\\">&nbsp;</td><td class=\\"buttons\\"><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base custom button elements 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><div><button class=\\"btn btn-danger\\">REMOVE</button></div></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><div><button class=\\"btn btn-danger\\">REMOVE</button></div></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><div><button class=\\"btn btn-danger\\">REMOVE</button></div></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><div><button class=\\"btn btn-danger\\">REMOVE</button></div></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><div><button class=\\"btn btn-danger\\">REMOVE</button></div></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><div><button class=\\"btn btn-danger\\">REMOVE</button></div></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base custom grid elements 1`] = `"<div><div class=\\"container-fluid bg-dark text-light\\"><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">product</div><div class=\\"col-lg-3 col-4 p-1\\">type</div><div class=\\"col-lg-2 col-5 p-1\\">price</div><div class=\\"col-lg-2 col-5 p-1\\">stock</div><div class=\\"col-lg-1 col-2 p-1\\"></div></div></div><div class=\\"container-fluid bg-light\\"><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">Desk</div><div class=\\"col-lg-3 col-4 p-1\\">capex</div><div class=\\"col-lg-2 col-5 p-1\\">100</div><div class=\\"col-lg-2 col-5 p-1\\">20</div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn border px-1 m-0 mx-1\\">x</button></div></div></div><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">Printer</div><div class=\\"col-lg-3 col-4 p-1\\">capex</div><div class=\\"col-lg-2 col-5 p-1\\">500</div><div class=\\"col-lg-2 col-5 p-1\\">10</div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn border px-1 m-0 mx-1\\">x</button></div></div></div><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">Paper</div><div class=\\"col-lg-3 col-4 p-1\\">consumable</div><div class=\\"col-lg-2 col-5 p-1\\">5</div><div class=\\"col-lg-2 col-5 p-1\\">2000</div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn border px-1 m-0 mx-1\\">x</button></div></div></div><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">Chair</div><div class=\\"col-lg-3 col-4 p-1\\">capex</div><div class=\\"col-lg-2 col-5 p-1\\">50</div><div class=\\"col-lg-2 col-5 p-1\\">50</div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn border px-1 m-0 mx-1\\">x</button></div></div></div><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">Computer</div><div class=\\"col-lg-3 col-4 p-1\\">capex</div><div class=\\"col-lg-2 col-5 p-1\\">1000</div><div class=\\"col-lg-2 col-5 p-1\\">20</div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn border px-1 m-0 mx-1\\">x</button></div></div></div><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">Rent</div><div class=\\"col-lg-3 col-4 p-1\\"></div><div class=\\"col-lg-2 col-5 p-1\\">2000</div><div class=\\"col-lg-2 col-5 p-1\\"></div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"><button class=\\"btn border px-1 m-0 mx-1\\">x</button></div></div></div><div class=\\"row\\"><div class=\\"col-lg-4 col-8 p-1\\">&nbsp;</div><div class=\\"col-lg-3 col-4 p-1\\">&nbsp;</div><div class=\\"col-lg-2 col-5 p-1\\">&nbsp;</div><div class=\\"col-lg-2 col-5 p-1\\">&nbsp;</div><div class=\\"col-lg-1 col-2 p-1\\"><div style=\\"display: flex; flex-direction: row;\\"></div></div></div></div></div>"`;

exports[`base externally triggered refresh 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base externally triggered refresh 2`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base no headers 1`] = `"<table><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base reference forwarding 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base single custom classes 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base single custom classes 2`] = `"<table><thead><tr class=\\"trClass\\"><th class=\\"thClass\\">product</th><th class=\\"thClass\\">type</th><th class=\\"thClass\\">price</th><th class=\\"thClass\\">stock</th><th class=\\"thClass\\"></th></tr></thead><tbody><tr class=\\"trClass\\"><td class=\\"tdClass\\">Desk</td><td class=\\"tdClass\\">capex</td><td class=\\"tdClass\\">100</td><td class=\\"tdClass\\">20</td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr class=\\"trClass\\"><td class=\\"tdClass\\">Printer</td><td class=\\"tdClass\\">capex</td><td class=\\"tdClass\\">500</td><td class=\\"tdClass\\">10</td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr class=\\"trClass\\"><td class=\\"tdClass\\">Paper</td><td class=\\"tdClass\\">consumable</td><td class=\\"tdClass\\">5</td><td class=\\"tdClass\\">2000</td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr class=\\"trClass\\"><td class=\\"tdClass\\">Chair</td><td class=\\"tdClass\\">capex</td><td class=\\"tdClass\\">50</td><td class=\\"tdClass\\">50</td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr class=\\"trClass\\"><td class=\\"tdClass\\">Computer</td><td class=\\"tdClass\\">capex</td><td class=\\"tdClass\\">1000</td><td class=\\"tdClass\\">20</td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr class=\\"trClass\\"><td class=\\"tdClass\\">Rent</td><td class=\\"tdClass\\"></td><td class=\\"tdClass\\">2000</td><td class=\\"tdClass\\"></td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr class=\\"trClass\\"><td class=\\"tdClass\\">&nbsp;</td><td class=\\"tdClass\\">&nbsp;</td><td class=\\"tdClass\\">&nbsp;</td><td class=\\"tdClass\\">&nbsp;</td><td class=\\"tdClass\\"><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;

exports[`base sync loader 1`] = `"<table><thead><tr><th>product</th><th>type</th><th>price</th><th>stock</th><th></th></tr></thead><tbody><tr><td>Desk</td><td>capex</td><td>100</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Printer</td><td>capex</td><td>500</td><td>10</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Paper</td><td>consumable</td><td>5</td><td>2000</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Chair</td><td>capex</td><td>50</td><td>50</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Computer</td><td>capex</td><td>1000</td><td>20</td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>Rent</td><td></td><td>2000</td><td></td><td><div style=\\"display: flex; flex-direction: row;\\"><button>x</button></div></td></tr><tr><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td>&nbsp;</td><td><div style=\\"display: flex; flex-direction: row;\\"></div></td></tr></tbody></table>"`;
